Sidecar qemu commandline args

victortoso opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

**What this PR does / why we need it**:
It adds another sidecar example which allow users to set custom QEMU command line arguments and even environment variables for QEMU.

**Which issue(s) this PR fixes**:
Fixes #8420 #7714

**Special notes for your reviewer**:
This sidecar is quite similar to SMBios one so I did implement the tests in the same file with minor refactors to existing code base.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

alicefr commented:
/cc

victortoso commented:
/retest

alicefr commented:
@victortoso the examples are automatically created, you are missing this part, therefore the `pull-kubevirt-generate` job fails and remove your example. Please, have a look to https://github.com/kubevirt/kubevirt/blob/main/tools/vms-generator/utils/utils.go#L53-LL79

fabiand commented:
@victortoso Nice.

Victor, I've always wondered if we can have a "meta sidecar" container to make it easier to write plugins.
Do you think you could change this container to become a base container with the grpc boilerplate, instead of having implementations for the hook points, this container would actually call out to a executable on the filesystem and provide the podspec and domxml as an input.
With this we could easily develop new custom hooks, by simply using this base container, and add a single script which covers one hook point, ie here:

```Dockerfile
FROM kubevirt-meta-hook
ADD ./inject-qemu-args.sh onDefineDomain
RUN chmod a+x onDefineDomain  # Just to be sure
```

IOW a hook would become a binary inside a container, where the binary name defines the hook point when it should be called.

I started to do this in https://github.com/kubevirt/kubevirt/compare/main...fabiand:kubevirt:shell-as-a-hook?expand=1 - but never finished

alicefr commented:
> @victortoso Nice.
> 
> Victor, I've always wondered if we can have a "meta sidecar" container to make it easier to write plugins. Do you think you could change this container to become a base container with the grpc boilerplate, instead of having implementations for the hook points, this container would actually call out to a executable on the filesystem and provide the podspec and domxml as an input. With this we could easily develop new custom hooks, by simply using this base container, and add a single script which covers one hook point, ie here:
> 
> ```dockerfile
> FROM kubevirt-meta-hook
> ADD ./inject-qemu-args.sh onDefineDomain
> RUN chmod a+x onDefineDomain  # Just to be sure
> ```
> 
> IOW a hook would become a binary inside a container, where the binary name defines the hook point when it should be called.
> 
> I started to do this in https://github.com/kubevirt/kubevirt/compare/main...fabiand:kubevirt:shell-as-a-hook?expand=1 - but never finished

@fabiand IMO, always writing an XML isn't very user-friendly, but probably better then always write a plugin from scratch. Additionally, you might want to customize specific parameters for the VM and if you provide the XML in the container image, you always need to rebuilt the image.

An alternative way to put the XML in the sidecar image could be a config map. A user could create the configmap with the custom XML and the sidecar could take it as input. However, this will require some changes in the sidecar API and virt-launcher as the configmap needs to be listed into the volumes of the virt-launcher pod and the volume mounted in the sidecar.

For a generic sidecar, I did this extension with virt-xml in the past: https://github.com/kubevirt/kubevirt/pull/6975. However, it required virt-xml to have access to the libvirt socket. Basically, the sidecar needs to have access to the empty dir where we store the libvirt socket. Again, this might be another alternative to the xml.

victortoso commented:
JFYI, this can be changed based on the discussion and results of https://github.com/kubevirt/kubevirt/pull/9231

victortoso commented:
/retest

alicefr commented on cmd/example-hook-sidecar-qemu-args/README.md:
````diff
@@ -0,0 +1,50 @@
+    <qemu:env name='G_MESSAGES_DEBUG' value='all'/>
+  </qemu:commandline>
+...
+```
````
@victortoso would be nice also to mention gdb as a use case and example of using this sidecar

victortoso replied:
I've added, let me know if you think I should tweak it further

alicefr commented on tests/vmi_hook_sidecar_test.go:
```diff
@@ -165,6 +171,127 @@ var _ = Describe("[sig-compute]HookSidecars", decorators.SigCompute, func() {
+	})
+
+	Describe("[vendor:cnv-qe@redhat.com][level:component] VMI definition", func() {
+		getVMIPod := func(vmi *v1.VirtualMachineInstance) (*k8sv1.Pod, bool, error) {
```
Can you rely on already existing functions like [GetRunningPodByVirtualMachineInstance](https://github.com/kubevirt/kubevirt/blob/main/tests/utils.go#L321)

victortoso replied:
For this function, we use [UnfinishedVMIPodSelector](https://github.com/kubevirt/kubevirt/blob/main/tests/utils.go#L1373)  as selector so I didn't find a helper function that uses it or accepts ` metav1.ListOptions` as parameter. We could create one as there seems to be enough calls to justify it
```bash
> grep -rniI "UnfinishedVMIPodSelector" tests/ | wc -l
13
```

alicefr commented on tests/vmi_hook_sidecar_test.go:
```diff
@@ -165,6 +171,127 @@ var _ = Describe("[sig-compute]HookSidecars", decorators.SigCompute, func() {
+
+				Eventually(func() bool {
+					vmiPod, exists, err := getVMIPod(vmi)
+					if err != nil {
```
You don't really need to check the err in the if statement if it is followed by `Expect(err).ToNot(HaveOccurred())`

victortoso replied:
Fixed, thanks!

victortoso commented:
/retest

victortoso commented:
I have rebased and changed the sidecar image to follow the proposal from https://github.com/kubevirt/kubevirt/pull/9229

victortoso commented:
/retest

Gohoy commented:
Could i use this feature now? I want to pass -s to the kubevirt vm to debug the kernel.

alicefr commented:
@Gohoy yes if you apply @victortoso patch, otherwise, you need to build your own image

victortoso commented:
> I'm happy to hear this answer, but I've never used sidecar, can I use this patch on older versions of kubevirt？

This sidecar uses the same API KubeVirt has for several versions of KubeVirt, so it shouldn't be a problem.

victortoso commented:
- Rebased
- Now using `sidecar-shim` base image introduced in https://github.com/kubevirt/kubevirt/pull/9231
- Moved under `cmd/sidecars/qemu-args`
- Renamed to `sidecar-qemu-args`

victortoso commented:
/retest-required

alicefr commented:
@victortoso this looks good to me. I only had some suggestion on the name and scope of the new image as we discussed in [slack](https://kubernetes.slack.com/archives/C0163DT0R8X/p1693468624646709). I'd love if others can give feedback on this.

/lgtm

victortoso commented:
> New changes are detected. LGTM label has been removed.

@alicefr applied the changes discussed which are only on the image name, in short:

```diff
 container_push(
-    name = "push-sidecar-qemu-args",
+    name = "push-debug-toolkit",
     format = "Docker",
-    image = "//cmd/sidecars/qemu-args:sidecar-qemu-args-image",
+    image = "//cmd/sidecars/qemu-args:debug-toolkit-image",
     registry = "$(container_prefix)",
-    repository = "$(image_prefix)sidecar-qemu-args",
+    repository = "$(image_prefix)debug-toolkit",
     tag = "$(container_tag)",
 )
```

and other places to reflect it. I've rebased the code as well, so I'm afraid the diff between versions could be annoying, sorry about that.

vladikr commented:
/lgtm
@victortoso thanks!

victortoso commented:
/test pull-kubevirt-check-tests-for-flakes

vasiliy-ul commented on cmd/sidecars/qemu-args/main.go:
```diff
@@ -0,0 +1,105 @@
+		domainSpec.QEMUCmd = &domainSchema.Commandline{}
+	}
+	// Set decoded annotations as QEMU Command line options
+	*domainSpec.QEMUCmd = api.Commandline(qemuCmd)
```
This will essentially overwrite the command line. KubeVirt may already have smth set here. Perhaps an `append` would be more appropriate, wdyt? But we need to ensure that the args/env are not re-added each time (i.e. they are added only once on the first hook call).

vasiliy-ul replied:
> But we need to ensure that the args/env are not re-added each time 

Ref: https://github.com/kubevirt/kubevirt/blob/3f7fa708521193396f759be35a3ad93cf8a8f1d0/cmd/network-slirp-binding/domain/domain.go#L85

vasiliy-ul replied:
>But we need to ensure that the args/env are not re-added each time

UPD: ... but after thinking a bit more and looking at the actual code, it seems that we cannot get the hook called twice on the same domain, so probably that part is not relevant. Or...?

victortoso replied:
I agree, appending is likely what we should do here to avoid removing settings from KubeVirt or other `sidecar` that has run prior to this one, if any.

vasiliy-ul commented on tests/vmi_hook_sidecar_test.go:
```diff
@@ -228,6 +234,123 @@ var _ = Describe("[sig-compute]HookSidecars", decorators.SigCompute, func() {
+
+			It("should successfully start with hook sidecar annotation for v1alpha2", func() {
+				By("Starting a VMI")
+				vmi, err = virtClient.VirtualMachineInstance(testsuite.GetTestNamespace(nil)).Create(context.Background(), vmi)
+				vmi.ObjectMeta.Annotations = RenderValidQEMUArgsSidecar(hooksv1alpha2.Version)
```
Probably you wanted to set the annotations first and then create the VMI.

victortoso replied:
:man_facepalming:

So, the test was using hooks version `v1alpha1` which makes this test the exactly same as the previous one.
I'll be fixing this. Thanks!

vasiliy-ul commented on tests/vmi_hook_sidecar_test.go:
```diff
@@ -228,6 +234,123 @@ var _ = Describe("[sig-compute]HookSidecars", decorators.SigCompute, func() {
+				Expect(err).ToNot(HaveOccurred())
+				logs := func() string { return getHookSidecarLogs(virtClient, vmi) }
+				libwait.WaitForSuccessfulVMIStart(vmi)
+				Eventually(logs,
+					11*time.Second,
+					500*time.Millisecond).
+					Should(ContainSubstring("Info method has been called"))
+				Eventually(logs,
+					11*time.Second,
+					500*time.Millisecond).
+					Should(ContainSubstring("OnDefineDomain method has been called"))
```
nit: I think you can combine the two:

```suggestion
				Eventually(logs,
					11*time.Second,
					500*time.Millisecond).
					Should(And(ContainSubstring("Info method has been called"), ContainSubstring("OnDefineDomain method has been called")))
```

victortoso replied:
Sure, I'll give it a try. Thanks!

victortoso commented:
The main changes were from @vasiliy-ul's review being the most noteworthy is the introduction of this [mergeQEMUCmd](https://github.com/victortoso/kubevirt/blob/sidecar-qemu-commandline-args/cmd/sidecars/qemu-args/main.go#L43) function, to merge with existing `custom` QEMU command lines that we might have from KubeVirt or other sidecars. It also logs if there is a clash.

victortoso commented:
/retest-required

victortoso commented:
/retest

victortoso commented:
/retest

victortoso commented:
Rebased.

Conflicts were from https://github.com/kubevirt/kubevirt/pull/10320 in `BUILD.bazel`, `hack/bazel-push-images.sh` and `hack/config-default.sh`, the last two due this commit d3b04fd125fdc418b47fbf432308a355d5a1ce73

We are very close to closing the loop. From this PR we created https://github.com/kubevirt/kubevirt/pull/9231 and then the follow-up https://github.com/kubevirt/kubevirt/pull/10320.

Thanks everyone from the comments, feedback and suggestions.

victortoso commented:
@vladikr, would be fine to get the lgtm [again](https://github.com/kubevirt/kubevirt/pull/9166#issuecomment-1701443061)?

vasiliy-ul commented on cmd/sidecars/qemu-args/main.go:
```diff
@@ -0,0 +1,155 @@
+
+	// Here we might warn if arg already exists.
+	for _, arg := range current.QEMUArg {
+		if _, exists := mapEnv[arg.Value]; exists {
```
Probably this should check in `mapArg`:

```suggestion
		if _, exists := mapArg[arg.Value]; exists {
```

victortoso commented:
@vasiliy-ul I've added some unit tests in https://github.com/kubevirt/kubevirt/commit/1d8e95086aa62a86a55a4477f65af981555b2a7e to be sure the merge function is doing the right thing. Thanks for the review!

vasiliy-ul commented on cmd/sidecars/qemu-args/main_test.go:
```diff
@@ -0,0 +1,111 @@
+			Expect(merged.QEMUArg).To(HaveLen(len(result.QEMUArg)))
+			Expect(merged.QEMUEnv).To(HaveLen(len(result.QEMUEnv)))
+			// With args, we kept the order intact as that might be important
+			Expect(reflect.DeepEqual(merged.QEMUArg, result.QEMUArg)).To(BeTrue())
```
nit: I think this can be simplified. Gomega already uses `reflect.DeepEqual` internally for checking equality https://onsi.github.io/gomega/#equalexpected-interface
```suggestion
			Expect(merged.QEMUArg).To(Equal(result.QEMUArg)
```

victortoso replied:
Yep. Didn't know that. Fixed!

vasiliy-ul reviewed (commented):
Looks good, thanks! 

/approve

victortoso commented:
I think this PR is not needed anymore. With https://github.com/kubevirt/kubevirt/pull/10479, users can use `sidecar-shim` container image with an extra script over ConfigMap to add commandline args. It would be left to the user/script to deal with [merging](https://github.com/kubevirt/kubevirt/pull/9166/commits/d9b41e04f869b54c9fe45fc8ee46a65dee6e2af9#diff-76e4b4b08de97fd8f502bd07da771fee543ed487392fe8da493bad26ac489891R43) command line args/envs but that's all.

victortoso commented:
I'll be closing this one in favor of using ConfigMap added by https://github.com/kubevirt/kubevirt/pull/10479
